crates/starknet/src/transactions.rs

```diff
@@ -219,14 +219,22 @@ impl StarknetTransaction {
         fn get_blockifier_messages_recursively(call_info: &CallInfo) -> Vec<OrderedMessageToL1> {
             let mut messages = vec![];
 
-            let from_address = call_info.call.caller_address.into();
+            // By default, `from_address` must correspond to the contract address that
+            // is sending the message. In the case of library calls, `code_address` is `None`,
+            // we then use the `caller_address` instead (which may be an account).
+            let from_address = if let Some(code_address) = call_info.call.code_address {
+                *code_address.0.key()
+            } else {
+                *call_info.call.caller_address.0.key()
+            };
 
             messages.extend(call_info.execution.l2_to_l1_messages.iter().map(|m| {
                 OrderedMessageToL1 {
                     order: m.order,
                     message: MessageToL1 {
                         to_address: m.message.to_address.into(),
-                        from_address,
+                        from_address: ContractAddress::new(from_address.into())
+                            .expect("Invalid contract address from blockifier's messsage"),
```

Comment:
@marioiordanov Do you have a better idea about this? Seems like changing `get_blockifier_messages_recursively` to return `Result` will only cause more changes

Reply:
right, it is used in two places, which will require cascade style changes. But I  dont see the method signature being changed

Reply:
I didn't mean that the code is already changing the method. I meant that the usual way of fixing this is by propagating the error, which would cause changing the type to `Result` and defining the needed errors (and first refactoring `.iter().map(...)` to a for-loop.

Reply:
yes, right

Reply:
It seems that if we use
```rust
            let from_address = if let Some(code_address) = call_info.call.code_address {
                code_address
            } else {
                call_info.call.caller_address
            };
```
then we can simply do
```rust
                    message: MessageToL1 {
                        ...,
                        from_address: from_address.into(),
                        ...
                    }
```

But I'm still taking a look at #298 which is addressing the same part of the code